splice_machine/src/main/java/com/splicemachine/derby/impl/sql/execute/actions/DependencyBucketing.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (c) 2012 - 2020 Splice Machine, Inc.
+ *
+ * This file is part of Splice Machine.
+ * Splice Machine is free software: you can redistribute it and/or modify it under the terms of the
+ * GNU Affero General Public License as published by the Free Software Foundation, either
+ * version 3, or (at your option) any later version.
+ * Splice Machine is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero General Public License for more details.
+ * You should have received a copy of the GNU Affero General Public License along with Splice Machine.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package com.splicemachine.derby.impl.sql.execute.actions;
+import splice.com.google.common.collect.*;
+
+import org.apache.log4j.Logger;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Helper class to sort objects to drop during drop schema cascade into buckets.
+ * We do not want to drop objects one by one, as we would need to inform other region servers
+ * every time we drop an object, so we want to drop them wave by wave. Each wave should contain
+ * objects that do not depend on each other.
+ * Objects that do not depend on anything can be part of the first wave.
+ */
+public class DependencyBucketing<T> extends DDLConstantOperation {
+    private Multimap<T, T> dependencies = ArrayListMultimap.create();
+    private Map<T, Integer> nodeLevels = new HashMap<>();
+    private int maxLevel = 1;
+
+    /**
+     * Add object to the dependency graph. If the object is not yet in the graph, it should
+     * be part of the first wave
+     * @param node
+     */
+    void addSingleNode(T node) {
+        nodeLevels.putIfAbsent(node, 0);
+    }
+
+    /**
+     * Add parent and child to the dependency graph. IF neither are in the graph when this function
+     * is called, they will respectively be part of the first and the second wave.
+     * If parent was already part of a certain bucket, child must go in the following one, and propagate
+     * that to its children.
+     * @param parent
+     * @param child
+     */
+    void addDependency(T parent, T child) {
+        dependencies.put(parent, child);
+        nodeLevels.putIfAbsent(parent, 0);
+        nodeLevels.putIfAbsent(child, 1);
+        propagateLevels(parent);
+    }
+
+    List<List<T>> getBuckets() {
+        List<List<T>> buckets = new ArrayList<>(maxLevel + 1);
+        for (int i = 0; i <= maxLevel; ++i) {
+            buckets.add(new ArrayList<>());
+        }
+        for (Map.Entry<T, Integer> level: nodeLevels.entrySet()) {
+            buckets.get(level.getValue()).add(level.getKey());
```

Comment:
Adding too many nodes increases the size of the PB message stored in ZK, ZooKeeper has [limits on the size](https://zookeeper.apache.org/doc/r3.1.2/api/org/apache/zookeeper/ZooKeeper.html) of the nodes:

>  The maximum allowable size of the data array is 1 MB (1,048,576 bytes). Arrays larger than this will cause a KeeperExecption to be thrown. 

If each table consumes 50 bytes then we would hit this limit with around 20K tables per schema which I don't think is unrealistic in production.

Please test this with larger tables, and think about chopping off the buckets into smaller sizes.